### PR TITLE
[gpt2pre 1.2] Start End Packer call method

### PR DESCRIPTION
**Pull request by pforderique**
Implements the full StartEndPacker class including it's `call()` method.
This is PR 2/2 of the full StartEndPacker implementation.

Depends on #7790.

NOTE:
The [keras implementation ](https://github.com/keras-team/keras-nlp/blob/master/keras_nlp/layers/start_end_packer.py#L48)includes a `return_padding_mask` parameter in the constructor. Setting this `True` makes the call method return the padded ragged tensor AND a mask of padded tokens as a tuple. This tuple return was not possible in the TypeScript implementation without changing the `call` type signature to `[Tensor|Tensor[], Tensor|Tensor[]]`. Therefore, I opted to create a separate method, `callAndReturnPaddingMask()` that returns just that. The `call` method simply takes the first result of this method call.

Room for optimization:
Currently, a mask will be calculated every time `call` is called, even when not needed. Decomposing the function further can help and can be done in the cleanup PR.

### CODE REVIEW

**Review by mattsoulanille — CHANGES_REQUESTED** [review]
I have a few suggestions. Nice test coverage!

File: tfjs-layers/src/layers/nlp/preprocessing/start_end_packer.ts [hunk 1]
```diff
@@ -88,11 +100,82 @@ export class StartEndPacker extends Layer {
...
+      }
+
+      const strInput = tensorToArr(input) as string[];
+
+      if (strInput.length <= length) {
+        const pads = Array(length - strInput.length).fill(padValue);
+        return tensor(strInput.concat(pads));
+      }
+
+      return tensor(strInput.slice(0, strInput.length - length));
```
[the remark this answers is not in the document]
Comment (pforderique) [reply tfjs-layers/src/layers/nlp/preprocessing/start_end_packer.ts:164]: Thanks for your review! 
This is true for when the user specifies to add an end token in which case we would never reach a point where `strInput.length > length`). However, in order to handle the case when the user only asks to prepend a start token (and not append an end token), we need this line since now the tensor shape is not guaranteed to be `<= sequenceLength`.

Because the function needs to account for this, the name might be misleading. I've changed it to `ensureLength` and added a test case to demonstrate the need for this line. Depending on how much it's used later in the NLP layers, `tf.ensureLength` could be a good op to support under [Transformations ](https://js.tensorflow.org/api/latest/#Tensors-Transformations) in the future.

Feel free to suggest other ways I could refactor this to make this more clear. Thanks!

File: tfjs-layers/src/layers/nlp/preprocessing/start_end_packer.ts [hunk 2]
```diff
@@ -88,11 +100,82 @@ export class StartEndPacker extends Layer {
...
+  }
+
+  /**
+   * Exactly like `call` except also returns a boolean padding mask of all
+   * locations that are filled in with the `padValue`.
+   */
+  callAndReturnPaddingMask(
+    inputs: Tensor|Tensor[],
+    kwargs: StartEndPackerOptions={addStartValue: true, addEndValue: true}
+  ): [Tensor|Tensor[], Tensor|Tensor[]] {
```
[the remark this answers is not in the document]
Comment (pforderique) [reply tfjs-layers/src/layers/nlp/preprocessing/start_end_packer.ts]: Nice catch! Changing all occurrences of `Tensor[]` to `Tensor2D` now that the tensor is padded helped simplify the test cases too :) The Keras implementation also allows for single tensor inputs (in which case you'll get a single tensor output). To keep this support, I changed the return types fo `Tensor1D|Tensor2D`. Let me know what you think. Thanks!

**Comment by pforderique** [comment]
@mattsoulanille your comments have been addressed. Thanks for the review!

**Review by mattsoulanille — APPROVED** [review]
LGTM

**Review by Linchenn — APPROVED** [review]
LGTM!
